aproxy.go

```diff
@@ -0,0 +1,404 @@
+		}
+	}()
+
+	conn.StartPreread()
+	defer conn.RestorePreread()
+	req, err := http.ReadRequest(bufio.NewReader(conn))
+	if err != nil {
+		return "", err
+	}
+	host := req.Host
+	if host == "" {
+		return "", errors.New("http request doesn't have host")
+	}
+	return host, nil
+}
+
+func DialProxy(proxy string) (net.Conn, error) {
+	proxyAddr, err := net.ResolveTCPAddr("tcp", proxy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve proxy address: %w", err)
+	}
+	conn, err := net.DialTCP("tcp", nil, proxyAddr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to proxy: %w", err)
+	}
+	return conn, nil
+}
+
+func DialProxyConnect(proxy string, dst string) (net.Conn, error) {
+	conn, err := DialProxy(proxy)
+	if err != nil {
+		return nil, err
+	}
+	request := http.Request{
+		Method: "CONNECT",
+		URL: &url.URL{
+			Host: dst,
+		},
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		Header: map[string][]string{
+			"User-Agent": {fmt.Sprintf("aproxy/%s", version)},
+		},
+		Host: dst,
+	}
+	err = request.Write(conn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send connect request to http proxy: %w", err)
+	}
+	response, err := http.ReadResponse(bufio.NewReaderSize(conn, 0), &request)
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("proxy return %d response for connect request", response.StatusCode)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to receive http connect response from proxy: %w", err)
+	}
+	return conn, nil
+}
+
+func GetOriginalDst(conn *net.TCPConn) (*net.TCPAddr, error) {
+	file, err := conn.File()
+	defer func(file *os.File) {
+		err := file.Close()
+		if err != nil {
+			slog.Error("failed to close the duplicated TCP socket file descriptor")
+		}
+	}(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert connection to file: %w", err)
+	}
+	var sockaddr [16]byte
+	size := 16
+	_, _, e := syscall.Syscall6(
+		syscall.SYS_GETSOCKOPT,
+		file.Fd(),
+		syscall.SOL_IP,
+		80, // SO_ORIGINAL_DST
+		uintptr(unsafe.Pointer(&sockaddr)),
+		uintptr(unsafe.Pointer(&size)),
+		0,
+	)
+	if e != 0 {
+		return nil, fmt.Errorf("getsockopt SO_ORIGINAL_DST failed: errno %d", e)
+	}
+	return &net.TCPAddr{
+		IP:   sockaddr[4:8],
+		Port: int(binary.BigEndian.Uint16(sockaddr[2:4])),
+	}, nil
+}
+
+func RelayTcp(conn io.ReadWriter, proxyConn io.ReadWriteCloser, logger *slog.Logger) {
+	var closed atomic.Bool
+	go func() {
+		_, err := io.Copy(proxyConn, conn)
+		if err != nil && !closed.Load() {
+			logger.Error("failed to relay network traffic to proxy", "error", err)
+		}
+		closed.Store(true)
+		_ = proxyConn.Close()
+	}()
+	_, err := io.Copy(conn, proxyConn)
+	if err != nil && !closed.Load() {
+		logger.Error("failed to relay network traffic from proxy", "error", err)
+	}
+	closed.Store(true)
+}
+
+func RelayHttp(conn io.ReadWriter, proxyConn io.ReadWriteCloser, logger *slog.Logger) {
+	defer func() {
+		_ = proxyConn.Close()
+	}()
```

Comment:
For things like this, `defer proxyConn.Close()` reads much nicer. Some linters don't like it, but I think that's on them, and that linter warning should be turned off. :-)

Reply:
Indeed, updated, thanks!